File: ml/linear_models.py
"""Linear models: LinearRegression, LogisticRegression."""

from __future__ import annotations

import numpy as np


class LinearRegression:
    def __init__(self, lr: float = 0.01, n_iterations: int = 1000) -> None:
        self.lr: float = lr
        self.n_iterations: int = n_iterations
        self.weights: np.ndarray | None = None
        self.bias: float | None = None

    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        n_samples, n_features = X.shape
        self.weights = np.zeros(n_features)
        self.bias = 0.0

        for _ in range(self.n_iterations):
            y_pred = np.dot(X, self.weights) + self.bias
            dw = (1 / n_samples) * np.dot(X.T, y_pred - y)
            db = (1 / n_samples) * np.sum(y_pred - y)
            self.weights -= self.lr * dw
            self.bias -= self.lr * db

    def predict(self, X: np.ndarray) -> np.ndarray:
        return np.dot(X, self.weights) + self.bias


class LogisticRegression:
    def __init__(self, lr: float = 0.01, n_iterations: int = 1000) -> None:
        self.lr: float = lr
        self.n_iterations: int = n_iterations
        self.weights: np.ndarray | None = None
        self.bias: float | None = None

    def _sigmoid(self, z: np.ndarray) -> np.ndarray:
        return 1 / (1 + np.exp(-np.clip(z, -500, 500)))

    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        n_samples, n_features = X.shape
        self.weights = np.zeros(n_features)
        self.bias = 0.0

        for _ in range(self.n_iterations):
            linear = np.dot(X, self.weights) + self.bias
            y_pred = self._sigmoid(linear)
            dw = (1 / n_samples) * np.dot(X.T, y_pred - y)
            db = (1 / n_samples) * np.sum(y_pred - y)
            self.weights -= self.lr * dw
            self.bias -= self.lr * db

    def predict(self, X: np.ndarray) -> np.ndarray:
        linear = np.dot(X, self.weights) + self.bias
        y_pred = self._sigmoid(linear)
        return (y_pred >= 0.5).astype(int)

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        linear = np.dot(X, self.weights) + self.bias
        y_pred = self._sigmoid(linear)
        return np.column_stack([1 - y_pred, y_pred])

File: ml/linear_models.ts
export class LinearRegression {
  weights: number[] = [];
  bias = 0;
  lr: number;
  n_iterations: number;

  constructor(lr = 0.01, n_iterations = 1000) {
    this.lr = lr;
    this.n_iterations = n_iterations;
  }

  fit(X: number[][], y: number[]): void {
    const m = X.length, n = X[0].length;
    this.weights = new Array(n).fill(0);
    this.bias = 0;
    for (let iter = 0; iter < this.n_iterations; iter++) {
      const pred = X.map(row => this.bias + this.weights.reduce((s, w, i) => s + w * row[i], 0));
      const error = pred.map((p, i) => p - y[i]);
      const dw = X[0].map((_, j) => this.lr * error.reduce((s, e, i) => s + e * X[i][j], 0) / m);
      this.bias -= this.lr * error.reduce((s, e) => s + e, 0) / m;
      this.weights = this.weights.map((w, j) => w - dw[j]);
    }
  }

  predict(X: number[][]): number[] {
    return X.map(row => this.bias + this.weights.reduce((s, w, i) => s + w * row[i], 0));
  }
}

export class LogisticRegression {
  weights: number[] = [];
  bias = 0;
  lr: number;
  n_iterations: number;

  constructor(lr = 0.1, n_iterations = 1000) {
    this.lr = lr;
    this.n_iterations = n_iterations;
  }

  sigmoid(z: number): number {
    return 1 / (1 + Math.exp(-Math.min(500, Math.max(-500, z))));
  }

  fit(X: number[][], y: number[]): void {
    const m = X.length, n = X[0].length;
    this.weights = new Array(n).fill(0);
    this.bias = 0;
    for (let iter = 0; iter < this.n_iterations; iter++) {
      const pred = X.map(row => this.sigmoid(this.bias + this.weights.reduce((s, w, i) => s + w * row[i], 0)));
      const error = pred.map((p, i) => p - y[i]);
      const dw = X[0].map((_, j) => this.lr * error.reduce((s, e, i) => s + e * X[i][j], 0) / m);
      this.bias -= this.lr * error.reduce((s, e) => s + e, 0) / m;
      this.weights = this.weights.map((w, j) => w - dw[j]);
    }
  }

  predict(X: number[][]): number[] {
    return X.map(row => (this.sigmoid(this.bias + this.weights.reduce((s, w, i) => s + w * row[i], 0)) > 0.5 ? 1 : 0));
  }

  predict_proba(X: number[][]): number[] {
    return X.map(row => this.sigmoid(this.bias + this.weights.reduce((s, w, i) => s + w * row[i], 0)));
  }
}